var/spack/repos/builtin/packages/gasnet/package.py

```diff
@@ -6,98 +6,133 @@
+    variant('udp', default=False, description="Portable UDP conduit, for Ethernet networks")
+
+    variant('debug', default=False, description="Enable library debugging mode")
+
+    depends_on('mpi', when='+mpi')
 
-        if '+pshm' in self.spec:
-            args.append('--enable-pshm')
+    depends_on('autoconf@2.69', type='build', when='@master:')
+    depends_on('automake@1.16:', type='build', when='@master:')
+
+    def conduits(self):
+        return [c for c in ['smp', 'mpi', 'ibv', 'udp'] if "+" + c in self.spec]
+
+    def install(self, spec, prefix):
+        if spec.satisfies('@master:'):
+            bootstrapsh = Executable("./Bootstrap")
+            bootstrapsh()
+
+        # The GASNet-EX library has a highly multi-dimensional configure space,
+        # to accomodate the varying behavioral requirements of each client runtime.
+        # The library's ABI/link compatibility is strongly dependent on these
+        # client-specific build-time settings, and that variability is deliberately NOT
+        # encoded in the variants of this package. The recommended way to build/deploy
+        # GASNet is as an EMBEDDED library within the build of the client package
+        # (eg. Berkeley UPC, UPC++, Legion, etc), some of which provide build-time
+        # selection of the GASNet library sources. This spack package provides
+        # the GASNet-EX sources, for use by appropriate client packages.
+        install_tree('.', prefix + "/src")
+
+        # Library build is provided for unit-testing purposes only (see notice above)
+        if len(self.conduits()) > 0:
+            options = ["--prefix=%s" % prefix]
+
+            if '+debug' in spec:
+                options.append("--enable-debug")
+
+            if '+mpi' in spec:
+                options.append("--enable-mpi-compat")
+            else:
+                options.append("--disable-mpi-compat")
+
+            options.append("--disable-auto-conduit-detect")
+            for c in self.conduits():
+                options.append("--enable-" + c)
+
+            configure(*options)
+            make()
+            make('install')
+
+            for c in self.conduits():
+                testdir = join_path(prefix, 'tests', c)
+                mkdirp(testdir)
+                make('-C', c + '-conduit', 'testgasnet-par')
+                install(c + "-conduit/testgasnet", testdir)
+            make('-C', c + '-conduit', 'testtools-par')
+            install(c + "-conduit/testtools", join_path(prefix, 'tests'))
+
+    @run_after('install')
+    @on_package_attributes(run_tests=True)
+    def test_install(self):
+        if '+smp' in self.spec:
+            make('-C', 'smp-conduit', 'run-tests')
+        if len(self.conduits()) > 0:
+            self.run_test(join_path(prefix, 'tests', 'testtools'),
+                          expected=['Done.'], status=0,
+                          installed=True, purpose="Running testtools")
```

Comment:
The output and status comparisons are the reason I wrote it this way. Otherwise a detected error seems more likely to go unnoticed. The test itself runs for about 20 seconds, so I'm not too worried about a small amount of parsing overhead.

On a related note, is there any way to have `run_test` look for _failure_ strings (eg the word "ERROR") in the output to trigger a test failure? The [documentation](https://spack.readthedocs.io/en/latest/packaging_guide.html#adding-stand-alone-tests) seems to imply this is possible but doesn't explain how.

Reply:
Did you try to provide it as expected output and it didn't work?  

Does the failure also return a non-zero status value?  If so, then you'll want to provide that as the expected status.  

If the failure could return different non-zero status values across versions -- or 0 for some versions and non-zero for others -- you could provide a list of "acceptable" status values.  We didn't get very explicit in this case in part because we weren't sure how often such a feature would be used.

Reply:
Let us know if you want to attempt this check in the PR or in a subsequent one.

Reply:
Hi @tldahlgren - I think you misunderstood my question. 

Our test prints messages (like "Done.") that indicate the test completed running. The test is also capable of printing messages (like "ERROR") that indicate a failure of a sub-test was detected and the test run should be considered a failure. In other words, the test's success condition is something like:  
```
(<test-output> contains <expected-string>) AND NOT (<test-output> contains <error-string>)
```

IIUC, the `run_test` interface handles the first condition, but it does not allow me to specify the latter (ie outputs that trigger test failure)..

I suspect this is a capability currently lacking in `run_test`, so I'm fine deferring this particular detail to a later PR.